src/adam7.rs

```diff
@@ -233,6 +238,98 @@ pub fn expand_pass(
     }
 }
 
+// TODO: Export this function in lib.rs.
+// Should we have a separate function, or can we use an enum to determine whether
+// to use default or splat?
+pub fn splat_expand_pass(
+    img: &mut [u8],
+    img_row_stride: usize,
+    interlaced_row: &[u8],
+    interlace_info: &Adam7Info,
+    bits_per_pixel: u8,
+) {
+    let bits_pp = bits_per_pixel as usize;
+
+    let bit_indices = expand_adam7_bits(img_row_stride, interlace_info, bits_pp);
+
+    if bits_pp < 8 {
+        for (pos, px) in bit_indices.zip(subbyte_pixels(interlaced_row, bits_pp)) {
+            bit_splat_expand(img, img_row_stride * 8, px, interlace_info, bits_pp, pos);
+        }
+    } else {
+        let bytes_pp = bits_pp / 8;
+
+        for (bitpos, px) in bit_indices.zip(interlaced_row.chunks(bytes_pp)) {
+            let start_byte = bitpos / 8;
+            byte_splat_expand(
+                img,
+                img_row_stride,
+                px,
+                interlace_info,
+                bytes_pp,
+                start_byte,
+            );
+        }
+    }
+}
+
+// TODO: Can we eliminate the duplicate code in |bit_splat_expand| and |byte_splat_expand|?
+// |stride| is bit-count here, unlike in |byte_splat_expand|, where it is byte-count.
+fn bit_splat_expand(
+    img: &mut [u8],
+    stride: usize,
+    px: u8,
+    info: &Adam7Info,
+    bits_pp: usize,
+    bit_pos: usize,
+) {
+    let height = ((img.len() * 8 - bit_pos) as f32 / stride as f32).ceil() as usize;
+    let (x_repeat, y_repeat) = SPLAT_EXPAND[info.pass as usize - 1];
+    for i in 0..min(y_repeat, height) {
+        let offset = bit_pos + i * stride;
+        let max_fill = min((stride - bit_pos % stride) / bits_pp, x_repeat);
+        copy_nbits_mtimes(img, offset, bits_pp, px, max_fill);
+    }
+}
+
+fn byte_splat_expand(
+    img: &mut [u8],
+    stride: usize,
+    px: &[u8],
+    info: &Adam7Info,
+    bytes_pp: usize,
+    start_byte: usize,
+) {
+    let height = ((img.len() - start_byte) as f32 / stride as f32).ceil() as usize;
+    let (x_repeat, y_repeat) = SPLAT_EXPAND[info.pass as usize - 1];
+    for i in 0..min(y_repeat, height) {
+        let offset = start_byte + i * stride;
+        let max_fill = min((stride - start_byte % stride) / bytes_pp, x_repeat);
+        copy_nbytes_mtimes(&mut img[offset..], px, bytes_pp, max_fill);
+    }
+}
+
+fn copy_nbits_mtimes(img: &mut [u8], bit_pos: usize, bits_pp: usize, px: u8, m: usize) {
```

Comment:
IIUC the implementation assumes that we never need to copy across **two** bytes.  So maybe the asserts below could be used to document this:

```
debug_assert_eq!(bit_pos % bits_pp, 0);
debug_assert!(bits_pp == 1 || bits_pp == 2 || bits_pp == 4);
```